File: C#/TreeNodeAdvState/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using Syncfusion.Windows.Forms.Tools;
using Syncfusion.Windows.Forms;

namespace TreeSearch
{
    public partial class Form1 : MetroForm
    {
        #region Declaration
        //To initialize the collection
        Dictionary<string, bool> dic = new Dictionary<string, bool>();
        #endregion

        #region Constructor
        public Form1()
        {
            InitializeComponent();
        }
        #endregion

        #region Implementation
        //To save TreeViewAdv.
        public Dictionary<string, bool> SaveTreeState(TreeViewAdv tree)
        {
            Dictionary<string, bool> nodeStates = new Dictionary<string, bool>();
          
            foreach (TreeNodeAdv node in Collect(tree.Nodes))
            {
                nodeStates.Add(node.Text, node.Expanded);
            }
            return nodeStates;
        }

        //To load TreeViewAdv
        private void RestoreTreeState(TreeViewAdv tree, Dictionary<string, bool> treeState)
        {
            foreach (TreeNodeAdv node in Collect(tree.Nodes))
            {
                if (treeState.Keys.Count != 0 && treeState[node.Text])
                    node.Expand();
                else
                    node.Expanded = false;
            }
        }

        //To iterate through all the nodes in TreeViewAdv
        IEnumerable<TreeNodeAdv> Collect(TreeNodeAdvCollection nodes)
        {
           foreach(TreeNodeAdv node in nodes)
           {
              yield return node;

              foreach (TreeNodeAdv child in Collect(node.Nodes))
              yield return child;
           }
        }
        #endregion

        #region Events
        private void buttonAdv1_Click(object sender, EventArgs e)
        {
            //To save the Tree State
            dic = SaveTreeState(this.treeViewAdv1);
        }

        private void buttonAdv2_Click(object sender, EventArgs e)
        {
            //To load the Tree state
            this.RestoreTreeState(this.treeViewAdv1, dic);
        }
        #endregion
    }
}

File: VB/TreeNodeAdvState_VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports Syncfusion.Windows.Forms.Tools
Imports Syncfusion.Windows.Forms

Namespace TreeSearch
	Partial Public Class Form1
		Inherits MetroForm
		#Region "Declaration"
		'To initialize the collection
		Private dic As New Dictionary(Of String, Boolean)()
		#End Region

		#Region "Constructor"
		Public Sub New()
			InitializeComponent()
		End Sub
		#End Region

		#Region "Implementation"
		'To save TreeViewAdv.
		Public Function SaveTreeState(ByVal tree As TreeViewAdv) As Dictionary(Of String, Boolean)
			Dim nodeStates As New Dictionary(Of String, Boolean)()

			For Each node As TreeNodeAdv In Collect(tree.Nodes)
				nodeStates.Add(node.Text, node.Expanded)
			Next node
			Return nodeStates
		End Function

		'To load TreeViewAdv
		Private Sub RestoreTreeState(ByVal tree As TreeViewAdv, ByVal treeState As Dictionary(Of String, Boolean))
			For Each node As TreeNodeAdv In Collect(tree.Nodes)
				If treeState.Keys.Count <> 0 AndAlso treeState(node.Text) Then
					node.Expand()
				Else
					node.Expanded = False
				End If
			Next node
		End Sub

		'To iterate through all the nodes in TreeViewAdv
		Private Iterator Function Collect(ByVal nodes As TreeNodeAdvCollection) As IEnumerable(Of TreeNodeAdv)
			For Each node As TreeNodeAdv In nodes
				Yield node

				For Each child As TreeNodeAdv In Collect(node.Nodes)
					Yield child
				Next child
			Next node
		End Function
#End Region

#Region "Events"
		Private Sub buttonAdv1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonAdv1.Click
			'To save the Tree State
			dic = SaveTreeState(Me.treeViewAdv1)
		End Sub

		Private Sub buttonAdv2_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonAdv2.Click
			'To load the Tree state
			Me.RestoreTreeState(Me.treeViewAdv1, dic)
		End Sub
		#End Region
	End Class
End Namespace
